backend/src/main/java/fk/prof/backend/model/association/impl/ZookeeperBasedBackendAssociationStore.java

```diff
@@ -0,0 +1,358 @@
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class ZookeeperBasedBackendAssociationStore implements BackendAssociationStore {
+  private static Logger logger = LoggerFactory.getLogger(ZookeeperBasedBackendAssociationStore.class);
+
+  private final Vertx vertx;
+  private final CuratorFramework curatorClient;
+  private final String backendAssociationPath;
+  private final int loadReportIntervalInSeconds;
+  private final int loadMissTolerance;
+
+  private final Map<String, BackendDetail> backendDetailLookup = new ConcurrentHashMap<>();
+  private final PriorityQueue<BackendDetail> availableBackendsByPriority;
+
+  private final Map<Recorder.ProcessGroup, String> processGroupToBackendLookup = new ConcurrentHashMap<>();
+  private final Map<Recorder.ProcessGroup, String> processGroupToZNodePathLookup = new ConcurrentHashMap<>();
+
+  private final ReentrantLock backendAssignmentLock = new ReentrantLock();
+
+  public ZookeeperBasedBackendAssociationStore(Vertx vertx, CuratorFramework curatorClient, String backendAssociationPath,
+                                                int loadReportIntervalInSeconds, int loadMissTolerance,
+                                                Comparator<BackendDetail> backendPriorityComparator)
+      throws Exception {
+    if(vertx == null) {
+      throw new IllegalStateException("Vertx instance is required");
+    }
+    if(curatorClient == null) {
+      throw new IllegalStateException("Curator client is required");
+    }
+    if(backendAssociationPath == null) {
+      throw new IllegalStateException("Backend association path in zookeeper hierarchy is required");
+    }
+    if(backendPriorityComparator == null) {
+      throw new IllegalStateException("Priority comparator for backends is required");
+    }
+
+    this.vertx = vertx;
+    this.curatorClient = curatorClient;
+    this.backendAssociationPath = backendAssociationPath;
+    this.loadReportIntervalInSeconds = loadReportIntervalInSeconds;
+    this.loadMissTolerance = loadMissTolerance;
+    this.availableBackendsByPriority = new PriorityQueue<>(backendPriorityComparator);
+
+    loadDataFromZookeeperInBackendLookup();
+    for(BackendDetail backendDetail: this.backendDetailLookup.values()) {
+      for(Recorder.ProcessGroup processGroup: backendDetail.getAssociatedProcessGroups()) {
+        if (this.processGroupToBackendLookup.putIfAbsent(processGroup, backendDetail.getBackendIPAddress()) != null) {
+          throw new IllegalStateException(String.format("Backend mapping already exists for process group=%s", processGroup));
+        }
+      }
+    }
+  }
+
+  @Override
+  public Future<Recorder.ProcessGroups> reportBackendLoad(BackendDTO.LoadReportRequest payload) {
+    String backendIPAddress = payload.getIp();
+    Future<Recorder.ProcessGroups> result = Future.future();
+    vertx.executeBlocking(future -> {
```

Comment:
A fast-path is possible for a known good backend (a backend that is not defunct and not new).

Its perhaps worth filing an issue for this and giving it a fast path in v2.